From e4b4675a34f1e1fc5d50c9720e9a34a7b952d15a Mon Sep 17 00:00:00 2001
From: Richard Chang <richardycc@google.com>
Date: Thu, 19 Feb 2026 16:05:06 +0000
Subject: [PATCH 18/63] ANDROID: zram: cache backing device block index for
 prefetched pages

When a page is prefetched from the backing device, zram currently
releases the associated backing device block immediately. This can lead
to race conditions if a page fault and prefetch occur simultaneously.

To address this, introduce a 'prefetch_cache' using an xarray to track
backing device block indices (blk_idx) for prefetched pages. Instead of
releasing the block in zram_populate_table(), the blk_idx is stored in
the prefetch_cache. The block is only released when
zram_slot_free_notify() is called, ensuring data validity on the backing
device as long as the swap slot is active.

This mechanism also enables an optimization: if a prefetched page is
selected for writeback again, zram can reuse the existing blk_idx from
the prefetch_cache, eliminating redundant write operations.

Additionally, skip recompression for prefetched slots to prevent
potential decompression algorithm mismatches during subsequent page
faults.

Bug: 495288884
Bug: 340137613
Change-Id: I8de78bb0a83465a1986f45074620166b482eb9b1
Signed-off-by: Richard Chang <richardycc@google.com>
(cherry picked from commit ed6a4b978dbd71cafdfc4a8ede13720488d80e6f)
---
 drivers/block/zram/zram_drv.c | 134 +++++++++++++++++++++++++++++++++-
 drivers/block/zram/zram_drv.h |  40 ++++++++++
 2 files changed, 171 insertions(+), 3 deletions(-)

diff --git a/drivers/block/zram/zram_drv.c b/drivers/block/zram/zram_drv.c
index 2155a8fafd4c..99f9ac79e483 100644
--- a/drivers/block/zram/zram_drv.c
+++ b/drivers/block/zram/zram_drv.c
@@ -644,6 +644,8 @@ static void reset_bdev(struct zram *zram)
 	zram->backing_dev = NULL;
 	zram->bdev = NULL;
 	zram->disk->fops = &zram_devops;
+
+	zram_prefetch_cache_destroy(zram);
 	kvfree(zram->bitmap);
 	zram->bitmap = NULL;
 }
@@ -999,11 +1001,99 @@ static struct zram_wb_req *zram_select_idle_req(struct zram_wb_ctl *wb_ctl)
 	return req;
 }
 
+#if IS_ENABLED(CONFIG_ZRAM_ANDROID_IOCTL)
+static int zram_prefetch_cache_pop(struct zram *zram, u32 index,
+				   unsigned long *blk_idx)
+{
+	void *val;
+
+	val = xa_erase(&zram->prefetch_cache, index);
+
+	if (!xa_is_value(val))
+		return -EINVAL;
+
+	*blk_idx = xa_to_value(val);
+	return 0;
+}
+
+void zram_prefetch_cache_init(struct zram *zram)
+{
+	xa_init(&zram->prefetch_cache);
+}
+
+void zram_prefetch_cache_destroy(struct zram *zram)
+{
+	xa_destroy(&zram->prefetch_cache);
+}
+
+bool zram_prefetch_cache_exist(struct zram *zram, u32 index)
+{
+	return xa_load(&zram->prefetch_cache, index) != NULL;
+}
+
+/* Return 1 on successful insertion, 0 when unsupported. and < 0 on error. */
+int zram_prefetch_cache_store(struct zram *zram, u32 index,
+			      unsigned long blk_idx)
+{
+	void *old_val;
+
+	/*
+	 * We use GFP_NOWAIT here to avoid the possible circular locking, and
+	 * prevent sleeping under the slot lock. The IO path might take the
+	 * fs_reclaim lock and zram_prefetch_cache_store is under the slot
+	 * lock.
+	 */
+	old_val = xa_store(&zram->prefetch_cache, index, xa_mk_value(blk_idx),
+			   GFP_NOWAIT);
+
+	return xa_is_err(old_val) ? xa_err(old_val) : 1;
+}
+
+/*
+ * If the slot was prefetched and is going to writeback again. We can reuse the
+ * blk_idx in prefetch_cache to reduce the extra write operations.
+ */
+int zram_prefetch_cache_reuse(struct zram *zram, u32 index)
+{
+	unsigned long blk_idx;
+	int err;
+
+	err = zram_prefetch_cache_pop(zram, index, &blk_idx);
+	if (err)
+		return err;
+
+	zram_clear_flag(zram, index, ZRAM_IDLE);
+	if (zram_test_flag(zram, index, ZRAM_HUGE))
+		atomic64_dec(&zram->stats.huge_pages);
+	atomic64_sub(zram_get_obj_size(zram, index),
+		     &zram->stats.compr_data_size);
+	zs_free(zram->mem_pool, zram_get_handle(zram, index));
+	zram_set_handle(zram, index, blk_idx);
+	zram_set_flag(zram, index, ZRAM_WB);
+
+	return 0;
+}
+
+int zram_prefetch_cache_drop(struct zram *zram, u32 index)
+{
+	unsigned long blk_idx;
+	int err;
+
+	err = zram_prefetch_cache_pop(zram, index, &blk_idx);
+	if (!err)
+		zram_release_bdev_block(zram, blk_idx);
+
+	return err;
+}
+#endif
+
 static int zram_populate_table(struct zram *zram, struct page *page, u32 index)
 {
 	unsigned long handle;
 	void *src, *dst;
 	u32 size;
+	int err;
+	unsigned long blk_idx;
 
 	zram_slot_lock(zram, index);
 	/*
@@ -1017,6 +1107,7 @@ static int zram_populate_table(struct zram *zram, struct page *page, u32 index)
 	}
 
 	size = zram_get_obj_size(zram, index);
+	blk_idx = zram_get_handle(zram, index);
 
 	handle = zs_malloc(zram->mem_pool, size,
 			   GFP_NOWAIT | __GFP_HIGHMEM |
@@ -1038,8 +1129,20 @@ static int zram_populate_table(struct zram *zram, struct page *page, u32 index)
 	zs_unmap_object(zram->mem_pool, handle);
 	kunmap_local(src);
 
+	/*
+	 * Retain blk_idx here and defer its release until
+	 * swap_slot_free_notify is triggered.
+	 */
+	err = zram_prefetch_cache_store(zram, index, blk_idx);
+	if (err < 0) {
+		zram_slot_unlock(zram, index);
+		zs_free(zram->mem_pool, handle);
+		return err;
+	} else if (err == 0) {
+		zram_release_bdev_block(zram, blk_idx);
+	}
+
 	zram_clear_flag(zram, index, ZRAM_WB);
-	zram_release_bdev_block(zram, zram_get_handle(zram, index));
 	zram_set_handle(zram, index, handle);
 	atomic64_add(size, &zram->stats.compr_data_size);
 	if (zram_test_flag(zram, index, ZRAM_HUGE))
@@ -1217,6 +1320,11 @@ int zram_writeback_slots(struct zram *zram,
 		 */
 		if (!zram_test_flag(zram, index, ZRAM_PP_SLOT))
 			goto next;
+
+		/* Reuse the blk_idx if it is found in the prefetch cache. */
+		if (zram_prefetch_cache_reuse(zram, index) == 0)
+			goto next;
+
 		if (zram->compressed_wb)
 			err = read_from_zspool_raw(zram, req->page, index);
 		else
@@ -1529,8 +1637,17 @@ static int decompress_bdev_page(struct zram *zram, struct page *page, u32 index)
 	void *src;
 
 	zram_slot_lock(zram, index);
-	/* Since slot was unlocked we need to make sure it's still ZRAM_WB */
-	if (!zram_test_flag(zram, index, ZRAM_WB)) {
+	/*
+	 * ZRAM_WB may have been cleared while the slot was unlocked, but
+	 * decompression remains safe. The upper swap layer guarantees that
+	 * swap_slot_free_notify will not free the slot during an active
+	 * page fault. Therefore, if ZRAM_WB was cleared, it must have been
+	 * by the prefetch path. Since prefetch uses prefetch_cache to keep
+	 * blk_idx valid until the slot is explicitly freed, blk_idx still
+	 * points to valid data for this fault.
+	 */
+	if (!zram_test_flag(zram, index, ZRAM_WB) &&
+	    !zram_prefetch_cache_exist(zram, index)) {
 		zram_slot_unlock(zram, index);
 		/* We read some stale data, zero it out */
 		memset_page(page, 0, 0, PAGE_SIZE);
@@ -2907,6 +3024,14 @@ static ssize_t recompress_store(struct device *dev,
 		if (!zram_test_flag(zram, pps->index, ZRAM_PP_SLOT))
 			goto next;
 
+		/*
+		 * Recompression of a prefetched slot could result in a page
+		 * fault by using the wrong decompression algorithm. So we skip
+		 * such slots during recompression.
+		 */
+		if (zram_prefetch_cache_exist(zram, pps->index))
+			goto next;
+
 		err = recompress_slot(zram, pps->index, page,
 				      &num_recomp_pages, threshold,
 				      prio, prio_max);
@@ -3068,6 +3193,7 @@ static void zram_slot_free_notify(struct block_device *bdev,
 	}
 
 	zram_free_page(zram, index);
+	zram_prefetch_cache_drop(zram, index);
 	zram_slot_unlock(zram, index);
 }
 
@@ -3151,6 +3277,8 @@ static ssize_t disksize_store(struct device *dev,
 		goto out_unlock;
 	}
 
+	zram_prefetch_cache_init(zram);
+
 	for (prio = 0; prio < ZRAM_MAX_COMPS; prio++) {
 		if (!zram->comp_algs[prio])
 			continue;
diff --git a/drivers/block/zram/zram_drv.h b/drivers/block/zram/zram_drv.h
index 67ef8be1df98..2a600c48baa1 100644
--- a/drivers/block/zram/zram_drv.h
+++ b/drivers/block/zram/zram_drv.h
@@ -19,6 +19,10 @@
 #include <linux/zsmalloc.h>
 #include <linux/crypto.h>
 
+#if IS_ENABLED(CONFIG_ZRAM_ANDROID_IOCTL)
+#include <linux/xarray.h>
+#endif
+
 #include "zcomp.h"
 
 #define SECTORS_PER_PAGE_SHIFT	(PAGE_SHIFT - SECTOR_SHIFT)
@@ -137,6 +141,9 @@ struct zram {
 	unsigned long *bitmap;
 	unsigned long nr_pages;
 #endif
+#if IS_ENABLED(CONFIG_ZRAM_ANDROID_IOCTL)
+	struct xarray prefetch_cache;
+#endif
 #ifdef CONFIG_ZRAM_MEMORY_TRACKING
 	struct dentry *debugfs_dir;
 #endif
@@ -187,4 +194,37 @@ int scan_slot_for_prefetch(struct zram *zram, unsigned long index,
 int zram_prefetch_slots(struct zram *zram, struct zram_pp_ctl *ctl);
 #endif
 
+#if IS_ENABLED(CONFIG_ZRAM_ANDROID_IOCTL)
+void zram_prefetch_cache_init(struct zram *zram);
+void zram_prefetch_cache_destroy(struct zram *zram);
+bool zram_prefetch_cache_exist(struct zram *zram, u32 index);
+int zram_prefetch_cache_store(struct zram *zram, u32 index,
+			      unsigned long blk_idx);
+int zram_prefetch_cache_reuse(struct zram *zram, u32 index);
+int zram_prefetch_cache_drop(struct zram *zram, u32 index);
+#else
+static inline void zram_prefetch_cache_init(struct zram *zram) {};
+static inline void zram_prefetch_cache_destroy(struct zram *zram) {};
+static inline bool zram_prefetch_cache_exist(struct zram *zram, u32 index)
+{
+	return false;
+}
+
+static inline int zram_prefetch_cache_store(struct zram *zram, u32 index,
+					    unsigned long blk_idx)
+{
+	return 0; /* unsupported */
+}
+
+static inline int zram_prefetch_cache_reuse(struct zram *zram, u32 index)
+{
+	return -EOPNOTSUPP;
+}
+
+static inline int zram_prefetch_cache_drop(struct zram *zram, u32 index)
+{
+	return -EOPNOTSUPP;
+}
+#endif
+
 #endif
-- 
2.50.1

